straightline.py

```diff
@@ -105,26 +119,81 @@ def process_image(self, image, screen):
         else:
-            # No markers found
-            self.stop_and_wait()
+            logger.info("no marker, looking for ribbon")
+            self.m_found = False
+            cropped_image = cv2.pyrDown(image, dstsize=(int(self.image_centre_x), int(self.image_centre_y)))
+            cropped_image = cropped_image[self.LINE_CROP_BOTTOM:self.LINE_CROP_TOP, self.LINE_CROP_LEFT:self.LINE_CROP_RIGHT]
+            img = cv2.cvtColor(cropped_image, cv2.COLOR_BGR2RGB)
+            ribbon_frame = pygame.surfarray.make_surface(cv2.flip(img, 1))
+            screen.blit(ribbon_frame, (0, 0))
+            blur_image = cv2.medianBlur(cropped_image, 3)
+            blur_image = cv2.cvtColor(blur_image, cv2.COLOR_RGB2HSV)
+            ribbon_mask = threshold_image(blur_image, self.ribbon_colour)
+            self.ribbon_following(ribbon_mask)
+            
         # Display the resulting frame
         frame = pygame.surfarray.make_surface(cv2.flip(frame,1))
-        screen.fill([0,0,0])
-        screen.blit(frame, (0,0))
+        screen.blit(frame, (self.LINE_CROP_TOP,0))
         pygame.display.update()
-        found_identifier = "F" if self.found else "NF"
-        img_name = "%d%simg.jpg" % (self.i, found_identifier)
+        m_found_identifier = "mF" if self.m_found else "NmF"
+        l_found_identifier = "lF" if self.l_found else "NlF"
+        img_name = "%d%s%s%dimg.jpg" % (self.i, m_found_identifier, l_found_identifier, self.ribbon_pos)
         # filesave for debugging: 
-        # cv2.imwrite(img_name, gray)
+#        if self.driving:
+#            cv2.imwrite(img_name, image)
         self.i += 1
 
+    def ribbon_following(self, ribbon_image):
+        ribbon_x, ribbon_y, ribbon_area, ribbon_contour = find_largest_contour(ribbon_image)
+        if ribbon_area > self.MIN_CONTOUR_AREA:
+            ribbon = [ribbon_x, ribbon_y, ribbon_area, ribbon_contour]
+            self.l_found = True
+            self.ribbon_pos = ribbon_x
+        else:
+            ribbon = None
+            self.l_found = False
+            self.ribbon_pos = 0
+        crop_width = self.LINE_CROP_RIGHT - self.LINE_CROP_LEFT
+        pygame.mouse.set_pos(int(ribbon_y), int(crop_width - ribbon_x))
+        self.follow_ribbon(ribbon, self.STRAIGHT_SPEED)
+
+    def follow_ribbon(self, ribbon, speed):
+        turn = 0.0
+        if ribbon is not None:
+            x = ribbon[0]
+            logger.info ("ribbon spotted at %i" % (x))
+            image_centre_x = (self.LINE_CROP_RIGHT - self.LINE_CROP_LEFT)/2
+            t_error  = float(image_centre_x - x) / image_centre_x
+            if self.aiming:
+                turn = self.AIM_P * t_error
+            else:
+                turn = self.LINE_TURN_P * t_error
+            turn = min(max(turn,-self.MAX_TURN_SPEED), self.MAX_TURN_SPEED)
+            if self.last_t_error is not None:
+                #if there was a real error last time then do some damping
+                if self.aiming:
```

Comment:
the turn factor is the only thing which differs here, so reduce repetition and opportunities for future bugs:
```
turn_factor = self.LINE_TURN_D if self.aiming else self.AIM_D
turn = turn - (turn_factor * (self.last_t_error - t_error))
```
note that I'm also removing the compound assignment operator in favour of the more explicit form. Personally I'm not a fan of them, as they're overly terse, so can make scanning the code slower, and also they can trip you up when dealing with lists (i know you're not here, but it's just a bit of a bad habit to get into)